src/java/org/apache/cassandra/io/tries/IncrementalTrieWriterPageAware.java

```diff
@@ -0,0 +1,715 @@
+    {
+        if (node.hasOutOfPageInBranch)
+        {
+            int sz = 0;
+            for (Node<Value> child : node.children)
+            {
+                if (depth < 64)
+                    sz += recalcTotalSizeRecursiveOnStack(child, nodePosition + sz, depth + 1);
+                else
+                    sz += recalcTotalSizeRecursiveOnHeap(child, nodePosition + sz);
+            }
+            node.branchSize = sz;
+        }
+
+        // The sizing below will use the branch size calculated above. Since that can change on out-of-page in branch,
+        // we need to recalculate the size if either flag is set.
+        if (node.hasOutOfPageChildren || node.hasOutOfPageInBranch)
+            node.nodeSize = serializer.sizeofNode(node, nodePosition + node.branchSize);
+
+        return node.branchSize + node.nodeSize;
+    }
+
+    private int recalcTotalSizeRecursiveOnHeap(Node<Value> node, long nodePosition) throws IOException
+    {
+        if (node.hasOutOfPageInBranch)
+            new RecalcTotalSizeRecursion(node, null, nodePosition).process();
+
+        if (node.hasOutOfPageChildren || node.hasOutOfPageInBranch)
+            node.nodeSize = serializer.sizeofNode(node, nodePosition + node.branchSize);
+
+        return node.branchSize + node.nodeSize;
+    }
+
+    class WriteRecursion extends Recursion<Node<Value>>
+    {
+        long nodePosition;
+
+        WriteRecursion(Node<Value> node, Recursion<Node<Value>> parent)
+        {
+            super(node, node.children.iterator(), parent);
+            nodePosition = dest.position();
+        }
+
+        @Override
+        Recursion<Node<Value>> makeChild(Node<Value> child)
+        {
+            if (child.filePos == -1)
+                return new WriteRecursion(child, this);
+            else
+                return null;
+        }
+
+        @Override
+        void complete() throws IOException
+        {
+            nodePosition = nodePosition + node.branchSize;
+            assert dest.position() == nodePosition
+                    : "Expected node position to be " + nodePosition + " but got " + dest.position() + " after writing children.\n" + dumpNode(node, dest.position());
+
+            serializer.write(dest, node, nodePosition);
+
+            assert dest.position() == nodePosition + node.nodeSize
+                    || dest.paddedPosition() == dest.position() // For PartitionIndexTest.testPointerGrowth where position may jump on page boundaries.
+                    : "Expected node position to be " + (nodePosition + node.nodeSize) + " but got " + dest.position() + " after writing node, nodeSize " + node.nodeSize + ".\n" + dumpNode(node, nodePosition);
+
+            node.filePos = nodePosition;
+        }
+    }
+
+    private long writeRecursive(Node<Value> node) throws IOException
+    {
+        return writeRecursiveOnStack(node, 0);
+    }
+
+    private long writeRecursiveOnStack(Node<Value> node, int depth) throws IOException
```

Comment:
hmm... I don't know if I understand it correctly - do you mean creating the whole trie writer subclass which calls super.xxx when depth < 64 and do other staff otherwise ?

Reply:
Specifically, I mean a version of the class [as it was before DB-2441 and DB-2364](https://github.com/blambov/bdp/blob/4a322e3d8ff8e44da31e78dd74592721c0fa1677/dse-db/src/java/org/apache/cassandra/io/tries/IncrementalTrieWriterPageAware.java) and a subclass that overrides the three recursive methods (and only them) to their current form without calling super's.

I agree this sounds like a lot of work for very unclear benefit, plus it adds a third copy of the implementation, but I do believe this is going to improve readability a lot. This is rather non-trivial piece of code and any improvement in that respect is worthwhile.

Reply:
I've just pushed a commit with this refactoring